File: examples/typescript/case04.test.ts
import assert from "node:assert/strict";
import { test } from "node:test";
import {
  ackBeforeEffect,
  applyEventOnce,
  handleRewardDelivery,
  newRewardStore,
  splitRewardAndMarker,
  weakEffectThenAck,
} from "./case04.ts";

test("delivery repeats but an atomic protected effect does not", () => {
  const weak = newRewardStore();
  const weakAck = { accepted: new Set<string>() };
  weakEffectThenAck(weak, weakAck, "E", "User42", 100, true);
  assert.equal(weakAck.accepted.has("E"), false);
  weakEffectThenAck(weak, weakAck, "E", "User42", 100, false);
  assert.equal(weak.state.balance.get("User42"), 200);

  const ackFirst = newRewardStore();
  const firstAck = { accepted: new Set<string>() };
  ackBeforeEffect(ackFirst, firstAck, "E", "User42", 100, true);
  assert.equal(firstAck.accepted.has("E"), true);
  assert.equal(ackFirst.state.balance.get("User42") ?? 0, 0);

  const strong = newRewardStore();
  assert.equal(applyEventOnce(strong, "E", "User42", 100), true);
  assert.equal(applyEventOnce(strong, "E", "User42", 100), false);
  assert.equal(strong.state.balance.get("User42"), 100);
  const strongAck = { accepted: new Set<string>() };
  handleRewardDelivery(strong, strongAck, "E", "User42", 100);
  assert.equal(strong.state.balance.get("User42"), 100);
  assert.equal(strongAck.accepted.has("E"), true);

  const split = newRewardStore();
  const marker = new Set<string>();
  splitRewardAndMarker(split, marker, "E", "User42", 100, true);
  splitRewardAndMarker(split, marker, "E", "User42", 100, false);
  assert.equal(split.state.balance.get("User42"), 200);
});

File: examples/python/case04_test.py
import unittest

from case04 import (
    DeliveryAck, RewardStore, ack_before_effect, apply_event_once,
    handle_reward_delivery, split_reward_and_marker, weak_effect_then_ack,
)


class Case04Test(unittest.TestCase):
    def test_delivery_and_protected_effect(self):
        weak, weak_ack = RewardStore(), DeliveryAck()
        weak_effect_then_ack(weak, weak_ack, "E", "User42", 100, True)
        self.assertNotIn("E", weak_ack.accepted)
        weak_effect_then_ack(weak, weak_ack, "E", "User42", 100, False)
        self.assertEqual(weak.balance["User42"], 200)

        ack_first, first_ack = RewardStore(), DeliveryAck()
        ack_before_effect(ack_first, first_ack, "E", "User42", 100, True)
        self.assertIn("E", first_ack.accepted)
        self.assertEqual(ack_first.balance.get("User42", 0), 0)

        strong, strong_ack = RewardStore(), DeliveryAck()
        self.assertTrue(apply_event_once(strong, "E", "User42", 100))
        self.assertFalse(apply_event_once(strong, "E", "User42", 100))
        self.assertEqual(strong.balance["User42"], 100)
        handle_reward_delivery(strong, strong_ack, "E", "User42", 100)
        self.assertEqual(strong.balance["User42"], 100)
        self.assertIn("E", strong_ack.accepted)

        split, marker = RewardStore(), set()
        split_reward_and_marker(split, marker, "E", "User42", 100, True)
        split_reward_and_marker(split, marker, "E", "User42", 100, False)
        self.assertEqual(split.balance["User42"], 200)


if __name__ == "__main__":
    unittest.main()
